resources/tentative/todomvc/src/components/todo-app/todo-app.component.js

```diff
@@ -0,0 +1,171 @@
+    }
+
+    updateItem(event) {
+        this._data.forEach((entry) => {
+            if (entry.id === event.detail.id)
+                entry.title = event.detail.title;
+
+        });
+
+        this.update("update-item", event.detail.id);
+    }
+
+    toggleItems(event) {
+        this.list.toggleItems(event.detail.completed);
+    }
+
+    clearCompletedItems() {
+        this.list.removeCompletedItems();
+    }
+
+    update(type = "", id = "") {
+        const totalItems = this._data.length;
+        const activeItems = this._data.filter((entry) => !entry.completed).length;
+        const completedItems = totalItems - activeItems;
+
+        this.list.setAttribute("total-items", totalItems);
+        this.list.updateElements(type, id);
+
+        this.topbar.setAttribute("total-items", totalItems);
+        this.topbar.setAttribute("active-items", activeItems);
+        this.topbar.setAttribute("completed-items", completedItems);
+
+        this.bottombar.setAttribute("total-items", totalItems);
+        this.bottombar.setAttribute("active-items", activeItems);
+    }
+
+    addListeners() {
+        this.topbar.addEventListener("toggle-all", this.toggleItems);
+        this.topbar.addEventListener("add-item", this.addItem);
+
+        this.list.addEventListener("toggle-item", this.toggleItem);
+        this.list.addEventListener("remove-item", this.removeItem);
+        this.list.addEventListener("update-item", this.updateItem);
+
+        this.bottombar.addEventListener(
+            "clear-completed-items",
+            this.clearCompletedItems
+        );
+    }
+
+    removeListeners() {
+        this.topbar.removeEventListener("toggle-all", this.toggleItems);
+        this.topbar.removeEventListener("add-item", this.addItem);
+
+        this.list.removeEventListener("toggle-item", this.toggleItem);
+        this.list.removeEventListener("remove-item", this.removeItem);
+        this.list.removeEventListener("update-item", this.updateItem);
+
+        this.bottombar.removeEventListener(
+            "clear-completed-items",
+            this.clearCompletedItems
+        );
+    }
+
+    routeChange = (route) => {
+        const routeName = route.split("/")[1] || "all";
+        this.list.updateRoute(routeName);
+        this.bottombar.updateRoute(routeName);
+        this.topbar.updateRoute(routeName);
+    };
+
+    connectedCallback() {
+        this.connected = true;
+        const node = document.importNode(template.content, true);
+
+        this.topbar = new TodoTopbar();
+        this.list = new TodoList();
+        this.bottombar = new TodoBottombar();
+
+        node.querySelector("[name=\"topbar\"]").append(this.topbar);
+        node.querySelector("[name=\"list\"]").append(this.list);
+        node.querySelector("[name=\"bottombar\"]").append(this.bottombar);
+
+        this.list.addItems(this._data);
+        this.addListeners();
+        this.update("connected");
+
+        this.shadow.append(node);
```

Comment:
Doesn't this mean we'd keep adding new clone of template each time the element becomes connected?
We should probably exit early when we've already done this once.

Reply:
yup - i'll adjust. This app doesn't re-connect at any point as far as I can tell, but it should be done better.

Reply:
changed approach and appending the template in the constructors. 
connectedCallback will now only update the template and add listeners